### PR TITLE
[Server42] LCM part 1 and 2

### PR DESCRIPTION
**Pull request by nikitamarchenko**
don't be scared of size; almost all (2,784) is auto-generated

### CODE REVIEW

File: src-java/server42/server42-control-storm-topology/src/main/java/org/openkilda/server42/control/topology/service/RouterService.java [hunk 1]
```diff
@@ -52,4 +58,29 @@ public void handleFlowRttFeatureToggle(Boolean featureEnabled) {
...
         }
     }
+
+    /**
+     * Part of LCM. Sends flow sync messages to FlowHandler or deactivate in case of feature disabled.
+     */
+    public void processSync() {
+        if (isFlowRttFeatureToggle()) {
+            Collection<SwitchProperties> all = switchPropertiesRepository.findAll();
+            all.stream()
```
Comment (surabujin) [inline src-java/server42/server42-control-storm-topology/src/main/java/org/openkilda/server42/control/topology/service/RouterService.java:68]: As for me simple loop will be better that filter stream 2 times.

**Review by surabujin — APPROVED** [review]

File: src-cpp/server42/src/FlowPoolTest.cpp [hunk 2]
```diff
@@ -1,82 +1,85 @@
...
+        flow.add_flow(ss.str(), BasicAllocator::allocate(i), i);
     }
 
     int **start = flow.table.data();
     int **end = flow.table.data() + flow.table.size();
 
-    for (int**pos = start;pos < end;pos += 64) {
+    for (int **pos = start; pos < end; pos += 64) {
         //m_Device->sendPackets(pos, std::min(64L, end - pos));
         std::cout << **pos << "-" << std::min(64L, end - pos) << "\n" << std::flush;
```
[the remark this answers is not in the document]
Comment (nikitamarchenko) [reply src-cpp/server42/src/FlowPoolTest.cpp:76]: no, ss is std::streamstring for string concatenation.

File: src-cpp/server42/src/PacketGenerator.cpp [hunk 3]
```diff
@@ -56,7 +66,17 @@ namespace org::openkilda {
 
         auto packet = flow_pool_t::allocator_t::allocate(newPacket.getRawPacket(), arg.device);
 
-        bool success = arg.flow_pool.add_flow(get_flow_id(arg), packet);
+        if (flow_meta) {
+            BOOST_LOG_TRIVIAL(debug)
+                << "remove old flow " << arg.flow_id
```
[the remark this answers is not in the document]
Comment (nikitamarchenko) [reply src-cpp/server42/src/PacketGenerator.cpp:71]: done

File: src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/Gate.java [hunk 4]
```diff
@@ -171,19 +204,34 @@ void listen(PushSettings data) {
...
         CommandPacket packet = builder.build();
-        try {
-            zeroMqClient.send(packet);
-        } catch (InvalidProtocolBufferException e) {
-            log.error("Marshalling error on {}", data);
-        }
+        zeroMqClient.send(packet);
+    }
+
+    private CommandPacket getFlowLIstCommandPacket(@Header(KafkaHeaders.RECEIVED_MESSAGE_KEY) String switchIdKey) {
```
[the remark this answers is not in the document]
Comment (nikitamarchenko) [reply src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/Gate.java:227]: done

File: src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/Gate.java [hunk 5]
```diff
@@ -154,6 +161,32 @@ void listen(ListFlowsRequest data) {
         }
     }
 
+    @KafkaHandler
+    void listen(ListFlowsOnSwitch data,
```
[the remark this answers is not in the document]
Comment (nikitamarchenko) [reply src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/Gate.java:165]: A prefer to use that command as a notification about a list of flow.

File: src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/Gate.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (nikitamarchenko) [reply src-java/server42/server42-control/src/main/java/org/openkilda/server42/control/kafka/Gate.java:227]: done

File: src-java/server42/server42-control/src/test/java/org/openkilda/server42/control/kafka/GateTest.java [hunk 6]
```diff
@@ -157,13 +157,23 @@ public void clearFlowsTest() throws Exception {
...
+        Any command = commandPacket.getCommand(0);
+        assertThat(command.is(Control.ClearFlowsFilter.class)).isTrue();
+
+        Control.ClearFlowsFilter unpack = command.unpack(Control.ClearFlowsFilter.class);
+        String dstMac = "1b:45:18:d6:71:5a";
+        assertThat(unpack.getDstMac()).isEqualTo(dstMac);
     }
 
 
+
```
[the remark this answers is not in the document]
Comment (nikitamarchenko) [reply src-java/server42/server42-control/src/test/java/org/openkilda/server42/control/kafka/GateTest.java:176]: done

File: src-java/server42/server42-control-storm-topology/src/main/java/org/openkilda/server42/control/topology/service/RouterService.java (hunk 1 above)
Comment (nikitamarchenko) [reply src-java/server42/server42-control-storm-topology/src/main/java/org/openkilda/server42/control/topology/service/RouterService.java:68]: It's a matter of taste.

File: src-cpp/server42/src/FlowId.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace org::openkilda {
+
+    enum class flow_id_members {
+        flow_id, direction
+    };
+
+    using flow_id_t = std::tuple<std::string, bool>;
+
+    inline flow_id_t get_flow_id(const std::string &flow_id, bool direction) {
```
[the remark this answers is not in the document]
Comment (nikitamarchenko) [reply src-cpp/server42/src/FlowId.h]: done

File: src-cpp/server42/src/FlowMetadata.h [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        int64_t hash;
+
+    public:
+        FlowMetadata(std::string flowId, bool direction, std::string dstMac, int64_t hash)
+                : dst_mac(std::move(dstMac)), hash(hash) {
+
+            flow_id = org::openkilda::get_flow_id(flowId, direction);
+        }
+
+        const std::string &get_flow_id() const {
```
[the remark this answers is not in the document]
Comment (nikitamarchenko) [reply src-cpp/server42/src/FlowMetadata.h:37]: done, I renamed to flow_endpoint

**Review by timofei-durakov — APPROVED** [review]
lgtm

File: src-cpp/server42/src/FlowMetadata.h [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+            flow_endpoint = org::openkilda::make_flow_endpoint(flow_id, direction);
+        }
+
+        const std::string &get_flow_id() const {
+
+            return std::get<int(flow_endpoint_members::flow_id)>(flow_endpoint);
+        };
+
+        const flow_endpoint_t &get_key() const {
```
Comment (surabujin) [inline src-cpp/server42/src/FlowMetadata.h:42]: shoud it become `get_flow_endpoint()`?

**Review by surabujin — APPROVED** [review]

File: src-cpp/server42/src/FlowMetadata.h (hunk 9 above)
Comment (nikitamarchenko) [reply src-cpp/server42/src/FlowMetadata.h]: done

**Pull request merged**